cli/src/commands/parse/pst.rs

```diff
@@ -0,0 +1,1275 @@
+
+    log::info!("Starting processing...");
+    for batch in &items_iter {
+        let mut emails = Vec::new();
+        for pst_message in batch {
+            match pst_message {
+                Ok(message) => match message.into_new_email(Mailbox(pst_file_name.clone())) {
+                    Ok(email) => emails.push(email),
+                    Err(e) => {
+                        if !args.resume_on_error {
+                            return Err(e);
+                        } else {
+                            statistics.add_failed_to_parse(1);
+                        }
+                    }
+                },
+                Err(e) => {
+                    if !args.resume_on_error {
+                        return Err(e);
+                    } else {
+                        statistics.add_failed_to_parse(1);
+                    }
+                }
+            }
+        }
+
+        let batch_len = emails.len();
+
+        if args.resume_on_error {
+            let result = client
+                .put_emails_split_on_failure(&bucket.full_name(), emails, args.no_charge)
+                .context("Could not upload batch of emails")?;
+            statistics.add_uploaded(batch_len - result.num_failed);
+            statistics.add_failed_to_upload(result.num_failed);
+        } else {
+            client.put_emails(&bucket.full_name(), emails, args.no_charge)?;
+            statistics.add_uploaded(batch_len);
+        };
+    }
+
+    if statistics.num_failed_to_parse() > 0 {
+        log::warn!(
+            "{} emails failed to parse.",
+            statistics.num_failed_to_parse()
+        )
+    }
+
+    if statistics.num_failed_to_upload() > 0 {
+        log::warn!(
+            "{} emails failed to upload.",
+            statistics.num_failed_to_upload()
+        )
+    }
+
+    log::info!(
+        "{} emails uploaded successfully.",
+        statistics.num_uploaded()
+    );
+
+    Ok(())
+}
+
+#[cfg(test)]
+mod tests {
+
+    use super::*;
+
+    #[test]
+    fn test_parse_pst() {
+        let pst_file =
+            PstFile::open(Path::new("tests/samples/test.pst")).expect("Could not open pst file");
+
+        let root_folder = pst_file
+            .get_root_folder()
+            .expect("Could not get test pst root folder");
+
+        let emails = root_folder
+            .all_items_iter()
+            .expect("Could not get all items iter")
+            .map(|message| {
+                message
+                    .expect("Could not parse message")
+                    .into_new_email(Mailbox("test.pst".to_string()))
+                    .expect("Could not create new email from pst message")
+            })
+            .collect::<Vec<NewEmail>>();
+
+        let expected_emails: Vec<NewEmail> =
+            serde_json::from_str(include_str!("../../../tests/samples/test.pst.json"))
+                .expect("Could not read expected emails");
+
+        assert_eq!(expected_emails, emails);
+    }
+}
```

Comment:
I'm a little suprised that this test (outputting data in JSONL) is not what the pst-extract type command does - generates you a file that you can then just upload using existing cli functionality.

This would separate the PST conversion and upload tasks, which I imagine probably both have their own kind of failure cases that it would be annoying to mix?

Reply:
Agree that it causes weird errors to mix however it follows the same pattern as `re parse eml` / `re parse msg` so prefer to be consistent with that

Reply:
That's fair enough